### PR TITLE
Add instructions for structured logging migration

### PR DESCRIPTION
**Pull request by serathius**
Ref https://github.com/kubernetes/enhancements/issues/1602
/cc @44past4 @brancz @DirectXMan12 @thockin

Wrote instructions for structured logging migration.

### CODE REVIEW

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+}
+```
+
+Log output before migration
+```
+I0525 17:27:16.857611  256278 proc.go:200] verb="GET" URI="/metrics" latency=1s resp=200 UserAgent="Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0. 2272.118 Safari/537.36." srcIP="127.0.0.1":
+```
+After
+```
+I0525 17:27:16.857697  256278 logtest.go:35] "HTTP" verb="GET" URI="/metrics" latency="1s" resp=200 userAgent="Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0. 2272.118 Safari/537.36." srcIP="127.0.0.1"
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:328]: I guess this should be proc.go

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+// output:
+// >> E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+func ErrorS(err error, msg string, keysAndValues ...interface{})
+
+// KObj is used to create ObjectRef when logging information about Kubernetes objects
+// Examples:
+// >> klog.InfoS("Pod status updated", "pod", klog.KObj(pod), "status", "ready")
+// output:
+// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+func KObj(obj metav1.Object) ObjectRef
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:59]: klog uses KMetadata not metav1.Object.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+}
+```
+
+## Migration
+
+1. Change log functions to structured equivalent
+1. Remove string formatting from log message
+1. Name arguments
+1. Use `klog.KObj` and `klog.KRef` for Kubernetes object references
+1. Compare outputted log
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:82]: This one is a bit unclear. I would change it to "Verify log output" or similar.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("HTTP", ...)
+}
+```
+
+### Name arguments
+
+Even thou new structured logging functions have very similar function prototype `func (string, ...interface{}) it's
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:170]: Missing `

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+small number of logs API calls. Based on criteria described in [selecting most important logs] section, selected 22 log
+calls are estimated to impact 99.9% of log volume. Up to date list of these log calls is provided in the
+[Enhancement Issue].
+
+[Enhancement Issue]: https://github.com/kubernetes/enhancements/issues/1602
+[selecting most important logs]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20191115-structured-logging.md#selecting-most-important-logs
+
+## Structured logging in Kubernetes
+
+With this enhancement a set of new functions were added to `klog`. As structured logging functions differ from greatly
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:22]: I am not sure what "differ from greatly" means in this context.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+```
+
+### Using ErrorS
+
+With `klog` structured logging borrowing interface from [logr] it also inherits it's differences in semantic of error
+function. This is visible in `klog.ErrorS` which requires passing an `error` as first argument. This is because in
+[logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:128]: Based on the klog implementation if nil is passed as error ErrorS implementation will act as InfoS.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+### Replacing Fatal calls
+
+Usage of Fatal should be discouraged and it's not available in new functions . Instead of depending on logger to exist
+process you should consider returning error and ensuring it is properly handled.
+
+## Remove string formatting from log message
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:135]: I believe that this should also cover verification of the log message and changing it when needed.

Apart from this some guidelines when it comes to proper logs messages would be probably appreciated. From simple things like if the massage should start with a capital letter and should be short to more complex ones like for instance the tense and form to be used: 
delete object -> object deleted
HTTP -> HTTP request received

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("HTTP", ...)
+}
+```
+
+### Name arguments
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:168]: I guess that even a short list of standard argument names like "status" could be helpful and would promote consistency.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+   klog.InfoS("Got request", "path", r.Path, "method", r.Method)
+}
+```
+
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different logs lines on `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:193]: What about the situation when the kind of an object is not known up front (for instance metav1.Object is passed as argument to function)? Should it be logged as "object" or rather the actual object kind should be used.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("updated pod", "pod", klog.KObj(pod))
+}
+```
+And
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("updated pod, uid: %s", pod.Uid)
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:233]: ObjectRef currently does not include uid so if we will replace uid with klog.KObj() we will have a regression when it comes to the information available in logs.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+}
+```
+
+### Compare results
+
+With introduction of structured functions log arguments will be formatted automatically instead of depending on the
+caller. This means that we can remove burden of picking format by caller and ensure greater log consistency, but during
+migration it's important to ensure that we avoid degradation of log quality. We should ensure that during migration we
+preserve properties like:
+* meaning
```
Comment (44past4) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:270]: What does "meaning" mean in this context?

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 6 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:128]: Should we mention this?

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 9 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:193]: Good question. I think best option would be to add new function to `klog` that would generate key based on `Kind`. But I would wait util we see some examples of this. For now I would recommend using "object"

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 10 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:233]: As UID and namespace name pair are both unique identifiers of object I would not treat this change as regression as it preserves uniqueness.

There is also case when name/namespaces are not available, then usage of UID is ok.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 11 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:270]: We should make sure that meaning of event that log is describing is preserved

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 6 above)
Comment (44past4) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:128]: As this is an implementation detail I would not mentioned it because someone may start relying on it. Instead I would state that this is forbiden.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 9 above)
Comment (44past4) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:193]: Agree.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 10 above)
Comment (44past4) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:233]: As I understand uid has a benefit of being unique across the time. So if a new resource will be created with the same name it will get a different uid.
I am not sure if there are any real use cases when this difference is relevant but I believe that this might be a regression for some users.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 8 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:183]: Do you have in mind listing arguments other then kubernetes kinds (those should be obvious)? I wanted to have this list evolve naturally during migration, instead of trying to force set of keys at this point. 

With log calls migrated it will be really easy to track global list of keys and propose refactors on whole repo.

Comment (44past4) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:183]: Yes, I was thinking about arguments like "status", "reason", "errorCode", "result" etc.
I am ok with creating such list latter on but I think that it would be good to have it somewhere in the documentation so that for common arguments we could have a proper names just from the start.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+1. Use `klog.KObj` and `klog.KRef` for Kubernetes object references
+1. Verify log output
+
+## Change log functions to structured equivalent
+
+Structured logging functions follow a different logging interface design than other functions in `klog`. They follow
+minimal design from [logr] thus there is no one-to-one mapping.
+
+Simplified mapping between functions:
+* `klog.Infof`, `klog.Info`, `klog.Infoln`, `klog.InfoDepth` -> `klog.InfoS`
```
Comment (thockin) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:98]: The "depth" ones are challenging because they presume hat popping up the call-stack is possible.  I count about 8 such files that make these calls in k/k.  Do you link we need an `InfoS` variant that includes depth?  That would be ugly to add to logr because it tries to abstract multiple implementations, but maybe that's something we should tackle in logr and klog, if it is a missing capability...

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+* `klog.V(N).Infof`, `klog.V(N).Info`, `klog.V(N).Infoln` -> `klog.V(N).InfoS`
+* `klog.Warning`, `klog.Warningf`, `klog.Warningln`, `klog.WarningDepth` -> `klog.InfoS`
+* `klog.Error`, `klog.Errorf`, `klog.Errorln`, `klog.ErrorDepth` -> `klog.ErrorS`
+* `klog.Fatal`, `klog.Fatalf`, `klog.Fatalln`, `klog.FatalDepth` -> `klog.ErrorS`
+
+### Removing Depth
+
+Functions with depth (`klog.InfoDepth`, `klog.WarningDepth`, `klog.ErrorDepth`, `klog.FatalDepth`) are used to indicate
+that the source of the log (added as metadata in log) is different than the invocation of logging library. This is
+usually used when implementing logging util functions. In those situations util functions should be removed and `klog`
```
Comment (thockin) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:108]: This is a difficult statement, I think.  Those util functions obviously exist for a reason.  Are we saying that there's NO valid reason to have a logging util and that there's really no recourse?

The example below is trivial.  A case like `staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go` is more interesting - maybe an example of how to convert something like that?

old:
```
func (rl *respLogger) Log() {
    latency := time.Since(rl.startTime)
    klog.InfoDepth(1, fmt.Sprintf("verb=%q URI=%q latency=%v UserAgent=%q srcIP=%q: hijacked",
        rl.req.Method, rl.req.RequestURI,
        latency, rl.req.UserAgent(), rl.req.RemoteAddr,
    ))  
}

// Called as:
foo.Log()
```

new:
```
func (rl *respLogger) GetLogPairs() []string {
    latency := time.Since(rl.startTime)
    return []string{
        "verb", rl.req.Method,
        "URI",rl.req.RequestURI
        "latency", latency,
        "UserAgent", rl.req.UserAgent(),
        "srcIP", rl.req.RemoteAddr,
        "hijacked", true,
    }
}

// Called as:
klog.InfoS("", foo.GetLogPairs()...)
```

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+error function. This is visible in `klog.ErrorS` which requires passing an `error` as the first argument. This is
+because in [logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating an error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
+
+### Replacing Fatal calls
+
+Use of Fatal should be discouraged and it's not available in new functions. Instead of depending on the logger to exit
+the process, you should consider returning an error and ensuring it is properly handled.
```
Comment (thockin) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:141]: If you are converting a site that assumes it must `Fatal()`, call `os.Exit()` yourself and document why.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 13 above)
Comment (thockin) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:108]: The point being that a helper is still possible and useful (duplicating that at every callsite would be bad) but it needs to be "turned inside out"

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different log lines of the `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
+names: `deployment`, `pod`, `node`, `replicaSet`. For objects of unknown type, is ok to log them under `object` key.
+
+In situations when we want to the log value of the same meaning twice (e.g. transition between state) it is ok to name
+an additional argument based on context, but leaving one main value with the name unchanged.
```
Comment (thockin) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:214]: if I have a bunch of lines that use "foo" as a key, and then I have one line that is changing foo, should I use {"foo", "newFoo"}, or {"oldFoo", "foo"} ?

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("Updated pod with uid: %s", pod.Uid)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("Updated pod", "pod", klog.KObj(pod))
```
Comment (thockin) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:257]: This loses the UID - is that intentional?  I assume it was added for a reason, and losing it will be a regression

Comment (DirectXMan12) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:257]: One of the nice things we can do with something like this is dynamically filter fields.  I'd suggest we capture UID, and then skip it at lower levels if someone doesn't want it.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 12 above)
Comment (thockin) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:98]: > Several of those are semi-global logging wrappers (1, 2, 5) that are better off just having an implementation that can do caller-skip (e.g. zap). I'm not sure we want the caller-skip at the interface level, though.

These ones seem like valid use-cases to me.  In fact, I wrote the first version of (5), I think.  The ability to strip frames from the callstack seems pretty necessary for any sort of logging util helper or adapter.

I don't quite grok your comment.  How would an implementation that can skip callframes be utilized under something like `logr` without a function for it in the interface?

In other words, should logr and klog grow new methods for `InfoDepth(depth, msg, string...)` and `ErrorDepth(depth, err, msg, string...)` ?

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 15 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:214]: Forgot to add this case. I'm leaning toward {"oldFoo", "foo"} if operation of changing foo succeeded and value under "foo" is the current. Some consequences of this:
* "foo" always points to most current/correct version 
* Little unnatural error handling . "foo" equal to `desiredFoo` on success, and "foo" equal to `oldFoo` if failed Example below:

```go
func ChangeFoo(desiredFoo, oldFoo) {
  err := changeFoo(desiredFoo, oldFoo)
  if err != nil {
    klog.ErrorS(err, "Failed changing foo", "desiredFoo", desiredFoo", "foo", oldFoo)
  }
  klog.InfoS("Changed foo", "oldFoo", oldFoo, "foo", desiredFoo)
}
```

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 16 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:257]: Proper handling UID was pushed outside scope of initial migration. Currently `klog.KObj` doesn't support UID. Dynamic filtering discussed by @DirectXMan12 is definitely on list of next steps. 

For now we have options:
* Use both `pod` and `pod_uid`. No regression, but can lead to duplication when `klog.KObj` starts using `UID`
* Replace usage of UID with `klog.KObj` by treating `<namespace>/<name>` as equivalent for `UID`. It would work in most places, but in some cases (when using `GenerateName`, or when recreating object where name stays same but UID changes) it will be regression. 

I think option 1 makes sense as it should be possible to scan code base and remove  for usage of `UID` before adding it to `klog.KObj`

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+```
+
+### Using ErrorS
+
+With `klog` structured logging borrowing the interface from [logr] it also inherits it's differences in semantic of
+error function. This is visible in `klog.ErrorS` which requires passing an `error` as the first argument. This is
+because in [logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating an error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
```
[the remark this answers is not in the document]
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:136]: Makes sense, I will apply your suggestion.

[the remark this answers is not in the document]
Comment (44past4) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:136]: It seems that the current implementation of klog.ErrorS() calls logr.Info() if the err is nil. If we want to encourage people to pass nil I believe that this should be changed so that logr.Error() would be called.

[the remark this answers is not in the document]
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:136]: Added https://github.com/kubernetes/klog/issues/153

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+}
+```
+
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different log lines of the `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
+names: `deployment`, `pod`, `node`, `replicaSet`. For objects of unknown type, is ok to log them under `object` key.
```
[the remark this answers is not in the document]
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:211]: sgtm

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 10 above)
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md]: Added more context in https://github.com/kubernetes/community/pull/4793#discussion_r431364645

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:136]: One note that `klog.ErrorS` is not guaranteed to generate a stacktrace as by default it still calls `klog.print`

[the remark this answers is not in the document]
Comment (DirectXMan12) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md]: Yep, we just kinda note in logr that `Error` may take special action to deal with errors in logr (see below).  For instance, I could see an impl that formatted errors for capture by special error recording tools (e.g. stackdriver error reporting or somesuch), ones that print stack traces (like zap), and so on.  It's nice because it leaves us a way to do cool stuff like that in the future, even if we don't do it by default now.

---

FWIW, logr docs say:

> Furthermore, certain implementations may choose to attach additional information (such as stack traces) on calls to Error, so it's preferred to use Error to log errors.

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 12 above)
Comment (DirectXMan12) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:98]: Hmm... let me rephrase a bit.  I think I misinterpreted somethings, and miscommunicated others.  I've rewritten this comment a few times, so it's possible there's something I'm not quite articulating correctly.

I agree that they're valid use cases, but:

All the mentioned numbers follow a pattern of "here's a set of helper functions where all the functions have a consistent caller skip".  The caller skip isn't varying per-function in the file.  I was originally (above) suggesting something like

```go
var logWrapper = klog.WithSkip(4)

func Info(msg) {
  logWrapper.InfoS(prefix+msg)
}
```

However, that only works when we're not redirecting to logr.  That's not too big a deal -- it preserves the status quo, punts on putting things into our interface, and should be doable with the way the internals of klog work.

We could consider adding the `WithSkip` to logr, which makes the caller skip compose and feels a lot more like the rest of logr's interface.  However, this is a bit annoying to implement in some logging backends (see below the fold), and still feels a bit brittle (again, below the fold).

<details>

<summary> [below the fold] Why I'm kinda against adding the extra Depth methods </summary>

My objections/uneasiness with this has two parts:

The first is that adding depth functions feels a lot like leaking implementation details through to the user.  This kinda paints us into the corner in terms of always explicitly supporting caller skip like this -- I could see a world where we decide to drop caller in favor of logger names + stack traces for errors or a world where we decide to do "capture caller into context".

The second is that that per-function caller skip is an easy interface to accidentally misuse:

- It's easy to miss/forget the `Depth` in one of your functions.

- It doesn't compose (you can't wrap helpers in helpers and still get useful information, which is something I've seen with similar types of functionality in codebases like controller-runtime).

- It's sufficiently brittle that I'd actively encourage folks not to use it -- it relies on exact code structure not changing, which isn't something we can enforce with the compiler, lint for, etc.  In the case of the etcd logger, if the internals of the etcd client change, we suddenly lose the usefulness of our logging information.  That's not really an interface I want people writing against.  It's also wildly unclear where the constant `4` comes from.

Moving on to practical matters: how much we care about people actually being able to implement this, or would we rather they just ignore it?  Zap doesn't support per-function caller skip OOTB, but it's doable with some hacking (you'd call `entry := logger.Check(...); entry.Caller = /* calculate caller yourself */`, or maybe could finagle a custom `zapcore.Core` to do it), logrus doesn't even *do* caller skip AFAICT, etc.  Do we want to continue to support delegating to logr?  It's not clear how much of a goal that it from the KEP (I think we should, or at least have some similar functionality).

</details>

<details>

<summary>I think we can do better here if/when we have Context</summary>

A similar idea whence to draw inspiration is of Rust's [track_caller](https://github.com/rust-lang/rfcs/blob/master/text/2091-inline-semantic.md#propagation-of-tracker).  While the original motivation wasn't quite the same as with Go, it has nice ergonomics: you stick a "#[track_caller]" on a function, and it gets propagated downward to any functions that also have `#[track_caller]`.  It composes nicely (since the caller tracking propagates downwards) and doesn't require explicit numbering for caller skip (see below the fold).

If we draw inspiration from the link above, we get an option that isn't quite doable with today's interface, but falls out naturally from the key words "propagates downwards": context.

```go
func MyHelper(ctx context.Context, msg string) {
    // save the caller if there's not one already, otherwise this becomes a no-op
    ctx = log.TrackCaller(ctx)
    klog.InfoSC(ctx, msg)
}

// also have a
// - log.ResetCaller, which resets the propogation.
// - log.BrittleTrackCallerWithSkip, which does what it says on the tin
```

Obvs this doesn't compose if the function can't take context, and requires context on the logger (a change to the interface).  However, it does mean that individual logging impls don't need to know how to grab caller information -- they just pull it out of the context or whatever.

Side benefit is that this also fixes the composability of things like `util/naming/from_stack.go`, which currently break when wrapped in helpers.

</details>

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+```
+
+[lowerCamelCase]: https://en.wiktionary.org/wiki/lowerCamelCase
+
+### Use `klog.KObj` and `klog.KRef` for Kubernetes objects
+
+As part of structured logging migration we want to ensure that kubernetes objects references are consistent within the
+codebase. Two new utility functions were introduced to klog `klog.KObj` and `klog.KRef`. Any reference
+(name, uid, namespace) to Kubernetes Object (Pod, Node, Deployment, CRD) should be rewritten to utilize those functions.
+In situations when object `UID` is would be beneficial for log, it should be added as separate field with `_uid` suffix.
```
Comment (DirectXMan12) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:244]: why not just have `UID` be a field in the output of `KObj` or have `KObjWithUID`?

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 20]
```diff
@@ -0,0 +1,375 @@
+# Structured Logging migration instructions
+
+This document describes instructions for migration proposed by [Structured Logging KEP]. It describes new structured
+functions introduced in `klog` (Kubernetes logging library) and how log calls should be changed to utilize new features.
+This document was written for the initial migration of `kubernetes/kubernetes` repository proposed for Alpha stage, but
+should be applicable at later stages or for other projects using `klog` logging library.
+
+[Structured Logging KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20191115-structured-logging.md
```
Comment (tahsinrahman) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:8]:
```suggestion
[Structured Logging KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging
```

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+
+## Goal of Alpha migration
+
+The first step is to introduce structure to the high percentage of logs generated in Kubernetes by changing only a
+small number of logs API calls. Based on criteria described in the [selecting most important logs] section, the selected
+22 log calls are estimated to impact 99.9% of log volume. The up to date list of these log calls is provided in the
+[Enhancement Issue].
+
+[Enhancement Issue]: https://github.com/kubernetes/enhancements/issues/1602
+[selecting most important logs]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20191115-structured-logging.md#selecting-most-important-logs
```
Comment (tahsinrahman) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:18]:
```suggestion
[selecting most important logs]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging#selecting-most-important-logs
```

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 22]
```diff
@@ -0,0 +1,375 @@
...
+
+## Structured logging in Kubernetes
+
+With this enhancement a set of new functions were added to `klog`. Structured logging functions the follow interface
+based on [logr], which hasa  different design than other `klog` functions which are based on [glog]. It is recommended
```
Comment (tahsinrahman) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:23]:
```suggestion
based on [logr], which has a  different design than other `klog` functions which are based on [glog]. It is recommended
```

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md [hunk 23]
```diff
@@ -0,0 +1,375 @@
...
+* `klog.Warning`, `klog.Warningf`, `klog.Warningln`, `klog.WarningDepth` -> `klog.InfoS`
+* `klog.Error`, `klog.Errorf`, `klog.Errorln`, `klog.ErrorDepth` -> `klog.ErrorS`
+* `klog.Fatal`, `klog.Fatalf`, `klog.Fatalln`, `klog.FatalDepth` -> `klog.ErrorS`
+
+### Removing Depth
+
+Functions with depth (`klog.InfoDepth`, `klog.WarningDepth`, `klog.ErrorDepth`, `klog.FatalDepth`) are used to indicate
+that the source of the log (added as metadata in log) is different than the invocation of logging library. This is
+usually used when implementing logging util functions. As logr interface doesn't support depth, those functions should 
+return logging arguments instead of calling `klog` directly.
```
Comment (DirectXMan12) [inline contributors/devel/sig-instrumentation/migration-to-structured-logging.md:109]: might be worth noting that there are situations where this isn't possible (mainly "adapters" and runtime.HandleError -- see my comment above), even if we decide to punt down the road on how to handle those cases

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 12 above)
Comment (thockin) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:98]: > We could consider adding the WithSkip to logr                                                                                                                 

The problem I have with this is that it doesn't feel like a property ofthe
logger object, but of the call-site.  I guess if you are wrapping a Logger in
some other type, you can argue that this IS part of the logger (or as you
characterized - it doesn't change).  If there's a one-off logging helper
function, you can always make a temporary object (though that can be
allocation-heav if it is done a lot).  E.g.

```
func logHelper(log logr.Logger, foo Foo) {
  logr.SkipFrames(1).Info("foo message", "address", foo.Addr(), "phone", foo.Phone)
}
```

> per-function caller skip is an easy interface to accidentally misuse:

OK, you convinced me.

> how much we care about people actually being able to implement this, or would we rather they just ignore it?

I suspect some log libs will handle it and some will no-op it.  I think that's
fine, honestly.  There will be pressure or there won't, but at least it is
_possible_ to DTRT

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 16 above)
Comment (thockin) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md]: I am OK with that

Comment (thockin) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md]: though Shouldn't it be `podUID` not `pod_uid` ?

Comment (serathius) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md]: Yes

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 12 above)
Comment (DirectXMan12) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:98]: > it can be alloc-heavy if it is done a lot

I would *hope* escape analysis and friends would prevent a heap allocation here.

**Comment by serathius** [comment]
ping @logicalhan

File: contributors/devel/sig-instrumentation/migration-to-structured-logging.md (hunk 23 above)
Comment (logicalhan) [reply contributors/devel/sig-instrumentation/migration-to-structured-logging.md:109]: maybe add a todo about expanding this section with the discussion above (we'll need it before beta)?
